File: scripts/render_diagrams.mjs
#!/usr/bin/env node

import { createHash } from "node:crypto";
import { mkdir, readFile, readdir, writeFile } from "node:fs/promises";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { instance } from "@viz-js/viz";

const root = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const sourceDirectory = join(root, "architecture", "source", "dot");
const packageMetadata = JSON.parse(await readFile(join(root, "package.json"), "utf8"));
const rendererPackageVersion = packageMetadata.devDependencies["@viz-js/viz"];
const expectedSources = [
  "components.dot",
  "containers.dot",
  "context.dot",
  "delegated-agent-sequence.dot",
  "evidence-flow.dot",
  "open-tier-sequence.dot",
  "protected-tier-sequence.dot",
  "six-control-spine.dot",
  "webmcp-sequence.dot",
];

function outputArgument(argv) {
  const index = argv.indexOf("--output");
  if (index === -1 || !argv[index + 1]) {
    throw new Error("Usage: render_diagrams.mjs --output <directory>");
  }
  return resolve(root, argv[index + 1]);
}

const outputDirectory = outputArgument(process.argv.slice(2));
const viz = await instance();
const sources = (await readdir(sourceDirectory))
  .filter((name) => name.endsWith(".dot"))
  .sort();

if (JSON.stringify(sources) !== JSON.stringify(expectedSources)) {
  throw new Error(
    `DOT source inventory mismatch: expected ${expectedSources.join(", ")}; ` +
      `found ${sources.join(", ") || "none"}`,
  );
}

await mkdir(outputDirectory, { recursive: true });
const manifest = [];

for (const name of sources) {
  const source = await readFile(join(sourceDirectory, name), "utf8");
  const result = viz.render(source, { engine: "dot", format: "svg" });
  if (result.status !== "success") {
    throw new Error(`${name}: ${result.errors.map((error) => error.message).join("; ")}`);
  }
  const svg = `${result.output.trimEnd()}\n`;
  const repeated = viz.render(source, { engine: "dot", format: "svg" });
  if (repeated.status !== "success" || `${repeated.output.trimEnd()}\n` !== svg) {
    throw new Error(`${name}: repeated render was not byte-for-byte deterministic`);
  }
  const outputName = name.replace(/\.dot$/, ".svg");
  await writeFile(join(outputDirectory, outputName), svg, "utf8");
  manifest.push({
    source: `architecture/source/dot/${name}`,
    output: outputName,
    sha256: createHash("sha256").update(svg).digest("hex"),
  });
}

const receipt = {
  renderer: "@viz-js/viz",
  rendererPackageVersion,
  graphvizVersion: viz.graphvizVersion,
  deterministicRepeat: true,
  diagrams: manifest,
};
await writeFile(
  join(outputDirectory, "manifest.json"),
  `${JSON.stringify(receipt, null, 2)}\n`,
  "utf8",
);
console.log(
  `Rendered ${manifest.length} diagrams with @viz-js/viz ${rendererPackageVersion} ` +
    `(Graphviz ${viz.graphvizVersion}); ` +
    `wrote checksums to ${join(outputDirectory, "manifest.json")}.`,
);

File: scripts/render_diagrams.py
#!/usr/bin/env python3
"""Render live DOT sources without modifying the immutable research pack."""

from __future__ import annotations

import argparse
import shutil
import subprocess
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()

    dot = shutil.which("dot")
    if dot is None:
        raise SystemExit("Graphviz 'dot' is required to render architecture diagrams")

    output = args.output if args.output.is_absolute() else ROOT / args.output
    output.mkdir(parents=True, exist_ok=True)
    sources = sorted((ROOT / "architecture" / "source" / "dot").glob("*.dot"))
    if len(sources) != 9:
        raise AssertionError(f"expected 9 DOT sources, found {len(sources)}")

    for source in sources:
        destination = output / f"{source.stem}.svg"
        subprocess.run([dot, "-Tsvg", str(source), "-o", str(destination)], check=True)
        if "<svg" not in destination.read_text(encoding="utf-8"):
            raise AssertionError(f"Graphviz did not create an SVG: {destination}")

    version = subprocess.run([dot, "-V"], capture_output=True, check=True, text=True)
    detail = (version.stderr or version.stdout).strip()
    print(f"Rendered {len(sources)} diagrams with {detail}.")


if __name__ == "__main__":
    main()
